File: 0450-delete-node-in-a-bst/0450-delete-node-in-a-bst.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def deleteNode(self, root: Optional[TreeNode], key: int) -> Optional[TreeNode]:
        if not root:
            return root
        
        if key < root.val:
            if root.left:
                root.left = self.deleteNode(root.left, key)
        elif key > root.val:
            if root.right:
                root.right = self.deleteNode(root.right, key)
        else:
            if not root.left and not root.right:
                return None
            elif not root.left:
                return root.right
            elif not root.right:
                return root.left
            
            min_val = root.right
            while min_val.left:
                min_val = min_val.left
            root.val = min_val.val
            root.right = self.deleteNode(root.right, root.val)
        return root
            
            

File: 0450-delete-node-in-a-bst/0450-delete-node-in-a-bst.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode deleteNode(TreeNode root, int key) {
        if(root == null) return null;
        if(key < root.val) {
            root.left = this.deleteNode(root.left, key);    
        } else if(key > root.val) {
            root.right = this.deleteNode(root.right, key);
        } else {
            if(root.left == null) {
                return root.right;
            } else if(root.right == null){
                return root.left;
            } else {
                TreeNode min = root.right;
                while(min.left != null) {
                    min = min.left;
                }
                root.val = min.val;
                root.right = this.deleteNode(root.right, min.val);   
            }
        }
        return root;
    }
}